src/protocol/malicious.rs

```diff
@@ -0,0 +1,306 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{fabric::Network, Direction},
+    protocol::{check_zero::check_zero, context::ProtocolContext, reveal::reveal, RecordId},
+    secret_sharing::Replicated,
+};
+use futures::future::try_join;
+
+use serde::{Deserialize, Serialize};
+
+/// A message sent by each helper when they've computed one share of the result
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct UValue<F> {
+    payload: F,
+}
+
+#[allow(dead_code)]
+pub struct SecurityValidator<'a, F, N> {
+    internally_indexed_ctx: ProtocolContext<'a, N>,
+    r_share: Replicated<F>,
+    u: F,
+    w: F,
+}
+
+impl<'a, F: Field, N: Network> SecurityValidator<'a, F, N> {
+    #[allow(dead_code)]
+    pub fn new(ctx: &ProtocolContext<'a, N>) -> SecurityValidator<'a, F, N> {
+        let internally_indexed_ctx = ctx.narrow(&"ValidationInternalIndexed".to_string());
+        let prss = internally_indexed_ctx.prss();
+
+        let r_share = prss.generate_replicated(RecordId::from(0));
+        let (u_left, u_right): (F, F) = prss.generate_fields(RecordId::from(1));
+        let (w_left, w_right): (F, F) = prss.generate_fields(RecordId::from(2));
+
+        SecurityValidator {
+            internally_indexed_ctx,
+            r_share,
+            u: u_right - u_left,
+            w: w_right - w_left,
+        }
+    }
+
+    #[allow(dead_code)]
+    fn compute_dot_product_contribution(a: Replicated<F>, b: Replicated<F>) -> F {
+        let (a_left, a_right) = a.as_tuple();
+        let (b_left, b_right) = b.as_tuple();
+        (a_left + a_right) * (b_left + b_right) - a_right * b_right
```

Comment:
Would this be easier?

```rust
(a.left() + a.right()) * (b.left() + b.right()) - a.right() * b.right()
```

It seems like most applications of "as_tuple()" would be more neatly expressed with "left()" and "right()" accessors.

Reply:
Yes, this would be better. I'll make that change.